File: Bezier/Bezier/BezierSpline.cpp
#include "stdafx.h"
#include "BezierSpline.h"
#include <iostream>

#define PI 3.14159265359

using namespace std;


BezierSpline::BezierSpline(vector<vector<double>> waypoints, vector<double>headings)
	: waypoints(waypoints)
	 ,headings(headings)
{
	convertHeadingToSlope();
	generateCubicBezier();
}

//Constructor for standard bezier that takes N degree and N+1 points for both x and y
BezierSpline::~BezierSpline()
{

}

vector<double> BezierSpline::calculateSplinePoint(double t)
{
	vector<double>point = piecewiseBeziers[(int)t].calculateBezierPoint(t-(double)((int)t));
	return point;
}

void BezierSpline::convertHeadingToSlope()
{
	for (int i = 0; i < headings.size(); i++)
	{
		if (headings[i] == 0.0)
			slopes.push_back(DBL_MAX-2000);
		else
			if (headings[i] == 180.0)
				slopes.push_back(DBL_MIN+2000);
		else
			slopes.push_back(1 / tan(headings[i] * PI / 180));
	}
}

void BezierSpline::generateCubicBezier()
{
	for (int i = 0; i < waypoints[0].size() - 1; i++) 
	{
		piecewiseBeziers.push_back(createCubicBezierSegment(waypoints[0][i], waypoints[1][i], waypoints[0][i+1], waypoints[1][i+1], slopes[i], slopes[i+1]));
		//cout << "Creating Cubic Bezier: " << endl << " x: " << waypoints[0][i] << endl << " y: " << waypoints[1][i] << endl << " x end: " << waypoints[0][i + 1] << endl << " y end: " << waypoints[1][i + 1] << endl << " slope: " << slopes[i] << endl << " slope end:" << slopes[i + 1] << endl;

	}
}

Bezier BezierSpline::createCubicBezierSegment(double startX, double startY, double endX, double endY, double ts, double te)
{
	cout << "STARTX: " << startX << endl;
	cout << "TS: " << ts << endl;
	cout << "TE: " << te << endl;
	cout << "ENDX: " << endX << endl;

	vector<double> x = { startX, (1.0/3.0)*1 + startX, endX - (1.0/3.0)*1, endX };
	vector<double> y = { startY, (1.0/3.0)*ts + startY, endY - (1.0/3.0)*te, endY };

	cout << "PO:" << startX << ", P1:" << (1.0 / 3.0)*1 + startX << ", P2:" << endX - (1.0 / 3.0)*1 << ", P3:" << endX << endl;
	cout << "PO:" << startY << ", P1:" << (1.0 / 3.0)*ts + startY << ", P2:" << endY - (1.0 / 3.0)*te << ", P3:" << endY << endl << endl;
	Bezier spline(x, y);
	return spline;
}

File: Python/bezier/bezierspline.py
import bezier as bz
import math

class BezierSpline:
  waypoints = []
  headings = []
  slopes = []
  piecewiseBeziers = []

  def __init__(self, way, head):
    self.waypoints = way
    self.headings = head
    self.convertHeadingToSlope()
    self.generateCubicBezier()

  def generateCubicBezier(self):
    for i in range(0, len(self.waypoints[0]) - 1):
      self.piecewiseBeziers.append(self.createCubicBezierSegment(self.waypoints[0][i], self.waypoints[1][i], self.waypoints[0][i+1], self.waypoints[1][i+1], self.slopes[i], self.slopes[i+1]))

  def calculateSplinePoint(self, t):
    point = self.piecewiseBeziers[int(t)].calculateBezierPoint(t - float(int(t)))
    return point

  def convertHeadingToSlope(self):
    ZERO = 0.00000001
    for i in range(0, len(self.headings)):
      if self.headings[i] == 0.0:
        self.slopes.append(ZERO)
      elif self.headings[i] == 180.0:
        self.slopes.append(ZERO)
      else:
        self.slopes.append(1 / math.tan(self.headings[i] * math.pi / 180.0))

  def createCubicBezierSegment(self, startX, startY, endX, endY, ts, te):
    # print('STARTX', startX)
    # print('TS', ts)
    # print('TE', te)
    # print('ENDX', endX)

    x = [startX, (1.0/3.0)*(endX - startX) + startX, endX - (1.0/3.0)*(endX - startX), endX]
    y = [startY, (1.0/3.0)*te + startY, endY - (1.0/3.0)*ts, endY]

    # print('P0', x[0], 'P1', x[1], 'P2', x[2], 'P3', x[3])
    # print('P0', y[0], 'P1', y[1], 'P2', y[2], 'P3', y[3])

    bez = bz.Bezier(x, y)
    return bez